cryptarchia/cryptarchia.py

```diff
@@ -218,5 +214,48 @@ def propose_block(self, slot: Slot, parent: BlockHeader) -> BlockHeader:
         return BlockHeader(parent=parent.id(), slot=slot)
 
 
+def common_prefix_len(a: Chain, b: Chain) -> int:
+    for i, (x, y) in enumerate(zip(a.blocks, b.blocks)):
+        if x.id != y.id:
+            return i
+    return min(len(a.blocks), len(b.blocks))
+
+
+def chain_density(chain: Chain, slot: Slot) -> int:
+    return len(
+        [
+            block
+            for block in chain.blocks
+            if block.slot.absolute_slot < slot.absolute_slot
+        ]
+    )
+
+
+# Implementation of the fork choice rule as defined in the Ouroboros Genesis paper
+# k defines the forking depth of chain we accept without more analysis
+# s defines the length of time after the fork happened we will inspect for chain density
+def maxvalid_bg(local_chain: Chain, forks: List[Chain], k: int, s: int) -> Chain:
+    cmax = local_chain
+    for chain in forks:
+        lowest_common_ancestor = common_prefix_len(cmax, chain)
+        m = cmax.length() - lowest_common_ancestor
+        if m <= k:
+            # Classic longest chain rule with parameter k
+            if cmax.length() < chain.length():
+                cmax = chain
+        else:
+            # The chain is forking too much, we need to pay a bit more attention
+            # In particular, select the chain that is the densest after the fork
+            forking_slot = Slot(
+                cmax.blocks[lowest_common_ancestor].slot.absolute_slot + s
+            )
+            cmax_density = chain_density(cmax, forking_slot)
+            candidate_density = chain_density(chain, forking_slot)
+            if cmax_density < candidate_density:
+                cmax = chain
```

Comment:
thinking out loud: If the second forking rule is so great, why do we need the first one? performance? 🤔

Reply:
I guess 's' can't be to small to avoid tails in the distribution of slot leaders, so you would need to wait a bit every time to observe the distribution after a fork. Also, if you were to propose a block after a slot with duplicate leaders you can't use a rule that needs information about future slots